### PR TITLE
Remove unused `# type: ignore`s

### PR DESCRIPTION
**Pull request by DMRobertson**
Over time we've begun to use newer versions of mypy, typeshed, stub
packages---and of course we've improved our own annotations. This makes
some type ignore comments no longer necessary. I have removed them.

There was one exception: a module that imports `select.epoll`. The
ignore is redundant on Linux, but I've kept it ignored for those of us
who work on the source tree using not-Linux. (#11771)

I'm more interested in the config line which enforces this. I want
unused ignores to be reported, because I think it's useful feedback when
annotating to know when you've fixed a problem you had to previously
ignore.

### CODE REVIEW

**Comment by DMRobertson** [comment]
mypy passes locally but fails in CI. That must be because I have `--extras all` installed locally, but not installed in CI.

Bugger. Maybe I should finally cave and pass through a list of extras to `setup-python-poetry`.

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -20,13 +20,9 @@ jobs:
...
-    # This does a vanilla `poetry install` - no extras. I'm slightly anxious
-    # that we might skip some typechecks on code that uses extras. However,
-    # I think the right way to fix this is to mark any extras needed for
-    # typechecking as development dependencies. To detect this, we ought to
-    # turn up mypy's strictness: disallow unknown imports and be accept fewer
-    # uses of `Any`.
-    uses: "matrix-org/backend-meta/.github/workflows/python-poetry-ci.yml@v1"
+    uses: "matrix-org/backend-meta/.github/workflows/python-poetry-ci.yml@dmr/extras-for-typechecking"
+    with:
+      typechecking-extras: "all"
```
Comment (DMRobertson) [inline .github/workflows/tests.yml:25]: Looks like I was right to be anxious.

On reflection, I think the approach I described on the left was too idealistic.

Before merging the branch of `backend-meta` should be reset to point to the v1 tag.

File: synapse/app/_base.py [hunk 2]
```diff
@@ -431,7 +431,7 @@ def run_sighup(*args: Any, **kwargs: Any) -> None:
     refresh_certificate(hs)
 
     # Start the tracer
-    synapse.logging.opentracing.init_tracer(hs)  # type: ignore[attr-defined] # noqa
+    init_tracer(hs)  # noqa
```
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/app/_base.py:434]: Was it something to do with circular imports maybe?

File: synapse/module_api/__init__.py [hunk 3]
```diff
@@ -1419,7 +1420,9 @@ def _validate_user_id(self, user_id: str) -> None:
                 f"{user_id} is not local to this homeserver; can't access account data for remote users."
             )
 
-    async def get_global(self, user_id: str, data_type: str) -> Optional[JsonDict]:
+    async def get_global(
+        self, user_id: str, data_type: str
+    ) -> Optional[Mapping[str, Any]]:
```
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/module_api/__init__.py:1425]: Can we use `JsonMapping` here, to convey json-serializability and read-only-ness?

**Comment by clokep** [comment]
> I'm more interested in the config line which enforces this. I want unused ignores to be reported, because I think it's useful feedback when annotating to know when you've fixed a problem you had to previously ignore.

Thanks for tackling this! 👍

**Comment by clokep** [comment]
I think @squahtx already gave some feedback on this, so I'm assuming the review is back to them!

**Review by squahtx — COMMENTED** [review]
It's great to see stale `# type: ignore`s being removed. I'm largely happy with the changes, save for the two questions/comments above.

**Comment by DMRobertson** [comment]
> It's great to see stale `# type: ignore`s being removed. I'm largely happy with the changes, save for the two questions/comments above.

Ahh sorry, I think I lost track of this today. I'll get those sorted tomorrow.

File: synapse/module_api/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/module_api/__init__.py:1425]: Sure. I think it should be fine to import this into the module api?

Really the module api should have an `__all__` to make it clear which symbols are considered publicly visible.

[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/module_api/__init__.py:1425]: Oh, it already does have `__all__`!

File: synapse/app/_base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/app/_base.py:434]: Not sure. Seems to stem from #5544, see https://github.com/matrix-org/synapse/pull/5544/files#diff-154d0c5dbc7a7bb36a8e89f20279bf1a51cdf77d1ca370482f95cd1ca644a6bbR246-R248

Prior to that change we were already running `import synapse` in that file, though I don't think we were using that import until the linked change.

TL;DR I don't think there's any particular reason.

**Review by squahtx — APPROVED** [review]
lgtm!

**Comment by DMRobertson** [comment]
Sytest failure seems to be https://github.com/matrix-org/sytest/issues/1113.
